### PR TITLE
Port timer and socket to use nanostack eventloop

### PR DESCRIPTION
**Pull request by kuggenhoffen**
- Bring in tick timer version of nanostack eventloop
- Port eventloop on top of pthread and posix timer
- Port M2MTimer to use eventloop
- Port M2MConnectionHandler to use eventloop for synchronised socket access

### CODE REVIEW

**Comment by kuggenhoffen** [comment]
@TeroJaasko @yogpan01 @anttiylitokola @simosillankorva @teetak01

**Comment by kuggenhoffen** [comment]
rebuild

**Comment by teetak01** [comment]
rebuild

**Comment by teetak01** [comment]
rebuild

File: source/m2mconnectionhandlerpimpl.cpp [hunk 1]
```diff
@@ -87,526 +175,589 @@ bool M2MConnectionHandlerPimpl::bind_connection(const uint16_t listen_port)
...
-                            if (connect(_socket_server, (const struct sockaddr *)&_sa_dst6, _slen_sa_dst6) != 0) {
-                                success = false;
-                                tr_error("M2MConnectionHandlerPimpl::resolve_hostname - failed to connect %s\n", ip_address);
-                            } else {
-                                success = true;
-                                tr_debug("M2MConnectionHandlerPimpl::resolve_hostname - connected to %s\n", ip_address);
-                            }
-                        }
-                        break;
-                }
```
Comment (TeroJaasko) [inline source/m2mconnectionhandlerpimpl.cpp]: This method might be const as it is not supposed to be modifying the object instance.

File: source/m2mconnectionhandlerpimpl.cpp [hunk 2]
```diff
@@ -87,526 +175,589 @@ bool M2MConnectionHandlerPimpl::bind_connection(const uint16_t listen_port)
...
-                            _received_packet_address->_length = sizeof(_resolved_address);
-
-                            _sa_dst6.sin6_family = AF_INET6;
-                            _sa_dst6.sin6_port = ntohs(server_port);
-                            memcpy(&_sa_dst6.sin6_addr,
-                                   _received_packet_address->_address,
-                                   _received_packet_address->_length);
-                            _sa_dst6.sin6_scope_id = 0;
-                            if (connect(_socket_server, (const struct sockaddr *)&_sa_dst6, _slen_sa_dst6) != 0) {
-                                success = false;
```
Comment (kuggenhoffen) [inline source/m2mconnectionhandlerpimpl.cpp]: There is strerror(errno) on next line in the new version, and your comment is pointing to the old stuff?

File: source/m2mconnectionhandlerpimpl.cpp (hunk 1 above)
Comment (kuggenhoffen) [inline source/m2mconnectionhandlerpimpl.cpp]: This is pointing to the old implementation again

**Comment by kuggenhoffen** [comment]
rebuild

**Comment by kuggenhoffen** [comment]
rebuild

**Comment by kuggenhoffen** [comment]
rebuild

**Comment by kuggenhoffen** [comment]
rebuild

**Comment by kuggenhoffen** [comment]
@yogpan01 @anttiylitokola @TeroJaasko 

Comments on this?

File: source/m2mconnectionhandlerpimpl.cpp [hunk 3]
```diff
@@ -13,71 +13,208 @@
...
-        _socket_server = -1;
+    fd_set sock_set;
+    // Check if we are in connecting state, if so wait until socket becomes writable
+    // and update socket state
+    while (_running && _socket >= 0 && _socket_state == ESocketStateConnecting) {
+        FD_ZERO(&sock_set);
+        // Add socket to read set
+        FD_SET(_socket, &sock_set);
+        if (select(_socket + 1, NULL, &sock_set, NULL, NULL) == -1) {
+            tr_debug("socket_listener() - write select fail!");
```
Comment (TeroJaasko) [inline source/m2mconnectionhandlerpimpl.cpp]: Shouldn't this be tr_error?

Comment (kuggenhoffen) [inline source/m2mconnectionhandlerpimpl.cpp]: I'll change that

File: source/m2mconnectionhandlerpimpl.cpp [hunk 4]
```diff
@@ -87,526 +224,710 @@ bool M2MConnectionHandlerPimpl::bind_connection(const uint16_t listen_port)
...
+            uint8_t* d = (uint8_t*)malloc(data_len+4);
 
-                if (rcv_size == -1) {
-                    tr_error("M2MConnectionHandlerPimpl::data_receive - error: %s", strerror(errno));
-                    _observer.socket_error(M2MConnectionHandler::SOCKET_READ_ERROR);
-                    _receive_data = false;
-                }
+            if(d){
+
+                d[0] = (data_len >> 24 )& 0xff;
```
Comment (TeroJaasko) [inline source/m2mconnectionhandlerpimpl.cpp]: data_len is 16b variable, at least the Coverity will complain for this.

**Comment by TeroJaasko** [comment]
I guess you need to add a couple of #includes to m2mconnectionhandlerpimpl.cpp as the mbed-client has removed some pointless includes. It would also be nice to see the code being reformatted to follow mbed style as it now varies quite a bit inside same files. 
Anyway, +1. This is definitely a big step forward on Linux version's reliability.

File: source/m2mconnectionhandlerpimpl.cpp [hunk 5]
```diff
@@ -87,526 +224,710 @@ bool M2MConnectionHandlerPimpl::bind_connection(const uint16_t listen_port)
...
+            } else {
+
+
+                if( recv_size < 4 ){
+
+                    close_socket();
+                    _observer.socket_error(M2MConnectionHandler::SOCKET_READ_ERROR, true);
+                    return;
+
+                }
```
Comment (kuggenhoffen) [inline source/m2mconnectionhandlerpimpl.cpp]: I'll make a new task for this, let's discuss further in there

**Comment by kuggenhoffen** [comment]
@TeroJaasko @anttiylitokola @yogpan01 
Ready to merge this?

**Pull request merged**